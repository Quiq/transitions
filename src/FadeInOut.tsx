import * as React from 'react';
import {Transition} from 'react-transition-group';
import styled, {css, keyframes} from 'react-emotion';
import {TransitionStatus} from 'react-transition-group/Transition';

export type VerticalSlideProps = {
  in: boolean;
  duration: number;
  children: React.ReactNode;
  className?: string;
};

const fadeIn = keyframes`
  0% {opacity: 0;}
  100% {opacity: 1;}
`;

const fadeOut = keyframes`
  0% {opacity: 1;}
  100% {opacity: 0;}
`;

interface FadeStyleProps {
  transition: TransitionStatus;
  duration: number;
}

const FadeContent = styled.div<FadeStyleProps>`
  ${props =>
    props.transition === 'entering' &&
    css`
      animation: ${fadeIn} ${props.duration}ms ease-in-out 0ms 1 normal both;
      overflow: hidden;
    `};

  ${props =>
    props.transition === 'exiting' &&
    css`
      animation: ${fadeOut} ${props.duration}ms ease-in-out 0ms 1 normal both;
      overflow: hidden;
    `};
`;

export const FadeInOut = ({in: inProp, duration, className, children}: VerticalSlideProps) => (
  <Transition in={inProp} timeout={duration} mountOnEnter unmountOnExit>
    {transition => (
      <FadeContent className={className} duration={duration} transition={transition}>
        {children}
      </FadeContent>
    )}
  </Transition>
);

export default FadeInOut;
