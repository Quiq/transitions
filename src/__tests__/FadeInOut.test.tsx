import React from 'react';
import FadeInOut from '../FadeInOut';
import {render, cleanup} from '@testing-library/react';
import * as emotion from 'emotion';
import {createSerializer} from 'jest-emotion';

expect.addSnapshotSerializer(createSerializer(emotion));

afterEach(cleanup);

jest.useFakeTimers();

const Content = () => <div>Test content</div>;

test('animating in', () => {
  const testProps = {
    duration: 100,
    children: <Content />,
  };
  const {container, rerender} = render(<FadeInOut in={false} {...testProps} />);

  rerender(<FadeInOut in {...testProps} />);

  expect(container).toMatchSnapshot();

  jest.runTimersToTime(100);
  expect(container).toMatchSnapshot();
});

test('animating out', () => {
  const testProps = {
    duration: 100,
    children: <Content />,
  };
  const {container, rerender} = render(<FadeInOut in={true} {...testProps} />);

  rerender(<FadeInOut in={false} {...testProps} />);

  expect(container).toMatchSnapshot();

  jest.runTimersToTime(100);
  expect(container).toMatchSnapshot();
});
