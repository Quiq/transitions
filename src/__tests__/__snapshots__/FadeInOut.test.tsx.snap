// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`animating in 1`] = `
.emotion-0 {
  -webkit-animation: animation-1o5jvb1 100ms ease-in-out 0ms 1 normal both;
  animation: animation-1o5jvb1 100ms ease-in-out 0ms 1 normal both;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
    data-testid="transitionContainer"
  >
    <div>
      Test content
    </div>
  </div>
</div>
`;

exports[`animating in 2`] = `
<div>
  <div
    class="css-3vl55h"
    data-testid="transitionContainer"
  >
    <div>
      Test content
    </div>
  </div>
</div>
`;

exports[`animating out 1`] = `
.emotion-0 {
  -webkit-animation: animation-1ilb1g4 100ms ease-in-out 0ms 1 normal both;
  animation: animation-1ilb1g4 100ms ease-in-out 0ms 1 normal both;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
    data-testid="transitionContainer"
  >
    <div>
      Test content
    </div>
  </div>
</div>
`;

exports[`animating out 2`] = `<div />`;
